src/box/memtx_tx.c

```diff
@@ -3369,16 +3473,31 @@ memtx_tx_nearby_gap_item_new(struct txn *txn, enum iterator_type type,
 }
 
 /**
- * Allocate and create count gap item.
+ * Allocate and create count gap item. The @a until tuple's story must have
+ * a gap item from the @a txn transaction or be tracked by it, so the story
+ * is not deleted by the garbage collector and the tuple is not deleted (if
+ * it's not NULL).
+ *
  * Note that in_read_gaps base member must be initialized later.
  */
 static struct count_gap_item *
 memtx_tx_count_gap_item_new(struct txn *txn, enum iterator_type type,
-			    const char *key, uint32_t part_count)
+			    const char *key, uint32_t part_count,
+			    struct tuple *until, hint_t until_hint)
 {
-	struct count_gap_item *item =
-		memtx_tx_nearby_gap_item_new(txn, type, key, part_count);
-	item->base.type = GAP_COUNT;
+	assert(until == NULL || tuple_has_flag(until, TUPLE_IS_DIRTY));
+
+	struct memtx_tx_mempool *pool = &txm.count_gap_item_mempool;
+	struct count_gap_item *item = memtx_tx_xmempool_alloc(txn, pool);
+	gap_item_base_create(&item->base, GAP_COUNT, txn);
+
+	item->type = type;
+	item->part_count = part_count;
+	item->key = memtx_tx_save_key(txn, key, part_count, item->short_key,
+				      sizeof(item->short_key), &item->key_len);
+	item->until = until;
```

Comment:
The tuple should be referenced here.

Reply:
Given the following statements:
1. The `until` tuple is [clarified](https://github.com/tarantool/tarantool/blob/3d984f6e0421c2442b396ddbd35ff7d4b1c1c8e6/src/box/memtx_tree.cc#L1022), so it's dirty and there's a story [hilding](https://github.com/tarantool/tarantool/blob/218fc46f5f44e6e27be73c6dceaa58f1eb377f46/src/box/memtx_tx.c#L967) it from the deletion.
   1. The tuple story has `GAP_INPLACE` item in its read_gaps.
2. And the story [won't](https://github.com/tarantool/tarantool/blob/218fc46f5f44e6e27be73c6dceaa58f1eb377f46/src/box/memtx_tx.c#L1627-L1632) be garbage collected until the last gap item that ues it is deleted.
4. The gap item will only be deleted after all gap items of the transaction are removed (including this count).

So if the count gap item is alive (and we can access its `until` member):
1. The transaction that created it is still alive.
2. The`until` tuple story is not garbage collected since it has the `GAP_INPLACE` of the transaction in it.
3. The tuple is not deleted since it has the story referencing it.

So it's OK to not reference the tuple here.

Reply:
Given the following statements:
1. The `until` tuple is [clarified](https://github.com/tarantool/tarantool/blob/3d984f6e0421c2442b396ddbd35ff7d4b1c1c8e6/src/box/memtx_tree.cc#L1022), so it's dirty and there's a stor
y [hilding](https://github.com/tarantool/tarantool/blob/218fc46f5f44e6e27be73c6dceaa58f1eb377f46/src/box/memtx_tx.c#L967) it from the deletion.
   1. The tuple story has `GAP_INPLACE` item in its read_gaps.
2. And the story [won't](https://github.com/tarantool/tarantool/blob/218fc46f5f44e6e27be73c6dceaa58f1eb377f46/src/box/memtx_tx.c#L1627-L1632) be garbage collected until the last gap item that ues it is deleted.
4. The gap item will only be deleted after all gap items of the transaction are [removed](https://github.com/tarantool/tarantool/blob/218fc46f5f44e6e27be73c6dceaa58f1eb377f46/src/box/memtx_tx.c#L3505-L3511) (including this count).

So if the count gap item is alive (and we can access its `until` member):
1. The transaction that created it is still alive.
2. The`until` tuple story is not garbage collected since it has the `GAP_INPLACE` of the transaction in it.
3. The tuple is not deleted since it has the story referencing it.

So it's OK to not reference the tuple here.

Reply:
When DDL happens, all the stories are deleted, including one referencing the tuple. So, even if the transaction is aborted by DDL, it's still actually alive (waits for rollback) so the count gap and the deleted tuple still can be accessed.

Reply:
Haven't considered the DDL machinery. Thanks!

```diff
--- a/src/box/memtx_tx.c
+++ b/src/box/memtx_tx.c
@@ -340,6 +340,12 @@ memtx_tx_count_gap_item_new(struct txn *txn, enum iterator_type type,
                            const char *key, uint32_t part_count,
                            struct tuple *until, hint_t until_hint);
 
+/**
+ * Destroy the count gap item.
+ */
+static void
+memtx_tx_count_gap_item_destroy(struct gap_item_base *base);
+
 /**
  * Helper structure for searching for point_hole_item in the hash table,
  * @sa point_hole_item_pool.
@@ -3165,6 +3165,7 @@ memtx_tx_delete_gap(struct gap_item_base *item)
                pool = &txm.nearby_gap_item_mempoool;
                break;
        case GAP_COUNT:
+               memtx_tx_count_gap_item_destroy(item);
                pool = &txm.count_gap_item_mempool;
                break;
        case GAP_FULL_SCAN:
@@ -3473,11 +3474,7 @@ memtx_tx_nearby_gap_item_new(struct txn *txn, enum iterator_type type,
 }
 
 /**
- * Allocate and create count gap item. The @a until tuple's story must have
- * a gap item from the @a txn transaction or be tracked by it, so the story
- * is not deleted by the garbage collector and the tuple is not deleted (if
- * it's not NULL).
- *
+ * Allocate and create count gap item.
  * Note that in_read_gaps base member must be initialized later.
  */
 static struct count_gap_item *
@@ -3497,10 +3494,20 @@ memtx_tx_count_gap_item_new(struct txn *txn, enum iterator_type type,
                                      sizeof(item->short_key), &item->key_len);
        item->until = until;
        item->until_hint = until_hint;
+       if (item->until != NULL)
+               tuple_ref(item->until);
 
        return item;
 }

+/**
+ * Destroy the count gap item.
+ */
+static void
+memtx_tx_count_gap_item_destroy(struct gap_item_base *base)
+{
+       struct count_gap_item *item = (struct count_gap_item *)base;
+       if (item->until != NULL)
+               tuple_unref(item->until);
+}
+
 /**
  * Allocate and create full scan gap item.
  * Note that in_read_gaps base member must be initialized later.
```

Reply:
@drewdzzz rebased and dropped these changes since the #10160 (which removes all the global gaps including `GAP_COUNT` on DDL along with stories) is merged.